src/backend/distributed/planner/distributed_planner.c

```diff
@@ -541,12 +541,25 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
  */
 static PlannedStmt *
 PlanFastPathDistributedStmt(DistributedPlanningContext *planContext,
-							Const *distributionKeyValue)
+							Node *distributionKeyValue)
 {
 	planContext->plannerRestrictionContext->fastPathRestrictionContext->
 	fastPathRouterQuery = true;
-	planContext->plannerRestrictionContext->fastPathRestrictionContext->
-	distributionKeyValue = distributionKeyValue;
+
+	if (distributionKeyValue == NULL)
+	{
+		/* nothing to record */
+	}
+	else if (IsA(distributionKeyValue, Const))
+	{
+		planContext->plannerRestrictionContext->fastPathRestrictionContext->
+		distributionKeyValue = (Const *) distributionKeyValue;
+	}
+	else if (IsA(distributionKeyValue, Param))
+	{
+		planContext->plannerRestrictionContext->fastPathRestrictionContext->
+		distributionKeyHasParam = true;
```

Comment:
Would be nice to unwrap this struct a little.

Reply:
done